packages/eslint-plugin-stories/src/rules/csf-v3-types.ts

```diff
@@ -0,0 +1,141 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';
+import isStories from '../utils/isStories';
+
+const failureMessages = {
+  specifyType:
+    'Specify a type of `StoryObj<Args>` for object stories, so that `composeStories` and Storybook can infer the correct args',
+};
+
+/**
+ * Enforce specifying an explicit type for story objects where necessary for `composeStories` and
+ * Storybook to correctly infer their args.
+ *
+ * For example, instead of
+ *
+ *   const Primary = {
+ *     render: (args) => { ... },
+ *   }
+ *
+ * this rules enforces adding a type annotation
+ *
+ *   const Primary: StoryObj<Args> = {
+ *     render: (args) => { ... },
+ *   }
+ *
+ * Only detects instances where we're sure a type annotation is needed. For example, when there's a
+ * render function defined that takes args (as in the examples above).
+ */
+const rule: TSESLint.RuleModule<keyof typeof failureMessages> = {
```

Comment:
Using `TSESLint.RuleModule` instead of the standard `Rule.RuleModule`.

Necessary to get good `@typescript-eslint/parser` types. They're compatible, but the `TSESLint` version enforces (via the type system) defining the rule in an oddly specific way.

For example, we have to define the optional `meta` property, and specify failure messages via `messageId` (instead of directly passing a message to `context.report`).

TL;DR - all this is valid eslint rule config, but it still looks a bit different than other rules in this plugin.